interfaces/builtin/docker.go

```diff
@@ -0,0 +1,642 @@
+sched_rr_get_interval
+sched_setaffinity
+sched_setattr
+sched_setparam
+sched_setscheduler
+sched_yield
+seccomp
+select
+semctl
+semget
+semop
+semtimedop
+send
+sendfile
+sendfile64
+sendmmsg
+sendmsg
+sendto
+setdomainname
+setfsgid
+setfsgid32
+setfsuid
+setfsuid32
+setgid
+setgid32
+setgroups
+setgroups32
+sethostname
+setitimer
+setns
+setpgid
+setpriority
+setregid
+setregid32
+setresgid
+setresgid32
+setresuid
+setresuid32
+setreuid
+setreuid32
+setrlimit
+set_robust_list
+setsid
+setsockopt
+set_thread_area
+set_tid_address
+settimeofday
+set_tls
+setuid
+setuid32
+setxattr
+shmat
+shmctl
+shmdt
+shmget
+shutdown
+sigaltstack
+signalfd
+signalfd4
+sigreturn
+socket
+socketcall
+socketpair
+splice
+stat
+stat64
+statfs
+statfs64
+stime
+symlink
+symlinkat
+sync
+sync_file_range
+syncfs
+sysinfo
+syslog
+tee
+tgkill
+time
+timer_create
+timer_delete
+timerfd_create
+timerfd_gettime
+timerfd_settime
+timer_getoverrun
+timer_gettime
+timer_settime
+times
+tkill
+truncate
+truncate64
+ugetrlimit
+umask
+umount
+umount2
+uname
+unlink
+unlinkat
+unshare
+utime
+utimensat
+utimes
+vfork
+vhangup
+vmsplice
+wait4
+waitid
+waitpid
+write
+writev
+`
+
+const dockerConnectedPlugAppArmor = `
+# Description: allow access to the Docker daemon socket. This gives privileged
+# access to the system via Docker's socket API.
+# Usage: reserved
+
+# Allow talking to the docker daemon
+/{,var/}run/docker.sock rw,
+`
+
+const dockerConnectedPlugSecComp = `
+# Description: allow access to the Docker daemon socket. This gives privileged
+# access to the system via Docker's socket API.
+# Usage: reserved
+
+setsockopt
+bind
+`
+
+const daemonPrivilegedAppArmor = `
+# Description: allow docker daemon to run containers invoked with
+# docker --privileged. This gives full access to all resources on the system
+# and thus gives device ownership to connected snaps.
+change_profile -> *,
+signal (send) peer=unconfined,
+ptrace (read, trace) peer=unconfined,
+/dev/** mrwkl,
+@{PROC}/** mrwkl,
+`
+
+const daemonPrivilegedSecComp = `
+# Description: allow docker daemon to run containers invoked with
+# docker --privileged. This gives full access to all resources on the system
+# and thus gives device ownership to connected snaps.
+@unrestricted
+`
+
+type DockerInterface struct{}
+
+func (iface *DockerInterface) Name() string {
+	return "docker"
+}
+
+func (iface *DockerInterface) PermanentPlugSnippet(plug *interfaces.Plug, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor,
+		interfaces.SecurityDBus,
+		interfaces.SecurityMount,
+		interfaces.SecuritySecComp,
+		interfaces.SecurityUDev:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *DockerInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	daemonPrivileged, _ := plug.Attrs["daemon-privileged"].(bool)
```

Comment:
This is a plug attribute to be used by the daemon not the cli command. Eg, consider:

```
plugs:
  docker-cli:
    interface: docker
  docker-privileged:
    interface: docker
    daemon-privileged: true
slots:
  docker-daemon:
    interface: docker

apps:
  docker:
    plugs:
    - docker-cli
    - network
    - home
  dockerd:
    daemon: simple
    plugs:
    - network-bind
    - firewall-control
    - docker-privileged
    slots:
    - docker-daemon
```

In this manner, dockerd uses 'daemon-privileged'. We've not done this before in interfaces but I believe it is consistent (I also tried to explain it in docs/interfaces.md). 'plugs: [ docker-privileged ]' isn't much different than if we wanted to do 'plugs: [ bluez ]' for dockerd where dockerd is allowed to connect to the bluez snap via the bluez interface except that in this case, it so happens that the docker interface is providing this. It also isn't much different than the 'plugs: [ network-bind ]' that we are already using, except that it is the docker interface that is providing the policy rather than the core snap. I did it this way because I felt it was important to keep the docker policy in one interface rather than pushing the privileged policy out to docker-privileged-support. If you'd prefer that, I can do it, but the user experience will end up equivalent with the snap connect command.